### PR TITLE
Add QANet to master branch.

### PR DESCRIPTION
**Pull request by qyhfbqz**
## Description ##
Generally, we have achieved the state-of-the-art performance (F1: 79.5 EM: 70.2) on SQuAD 1.1 dataset.

1. Add data preprocessing including specific bucketing, batchify, etc to be aligned with TF implementation.
2. Add QANet including character and word embedding layer, embedding block, encoder block(e.g. position encoder, depthwise conv, self-attention, positionwise ffn), output layer.
3. Add attention layer including the trilinear similarity computation, etc.
4. Add official evaluation script to report the F1 and EM score.
5. Add other utilities in the training including Exponential Moving Average, Warmup, etc.
6. Add [ConvolutionalEncoder](https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/model/convolutional_encoder.py#L34) to replace traditional 1-D convolution in char convolution layer.
7. Refine the hyperparameters by using argparse package.

### CODE REVIEW

**Comment by cgraywang** [comment]
Thanks @qyhfbqz for the hard work!

Let's now focus on reviewing the following files, mainly QANet model and the training script:

> scripts/question_answering/model.py
> scripts/question_answering/question_answering.rst
> scripts/question_answering/train.py
> scripts/question_answering/util.py

The data processing/data loader/official evaluation scripts will be improved as next steps.

Most discussions have been done in a [separate fork](https://github.com/cgraywang/gluon-nlp-1/pull/8) and in the previous [feature branch fork](https://github.com/dmlc/gluon-nlp/pull/339).

**Comment by mli** [comment]
Job PR-411/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-411/7/index.html

File: scripts/question_answering/index.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Reference: A. Yu, D. Dohan et al. "`QANet: Combining Local Convolution with Global Self-Attention for Reading Comprehension <https://arxiv.org/pdf/1804.09541.pdf>`_". ICLR 2018
+
+The dataset used for training and testing the models is `SQuAD <https://arxiv.org/pdf/1606.05250.pdf>`_ v1.0 dataset.
+
+The key features used to reproduce the results for pre-trained models are listed in the following tables.
+
++--------------------------------------+---------------+---------------+--------------------------------------------+--------------------------------------------+--------------+
+|                 Model                | QANet_MXNet_1 | QANet_MXNet_2 |                 QANet_tf_1                 |                 QANet_tf_2                 | origin_paper |
++--------------------------------------+---------------+---------------+--------------------------------------------+--------------------------------------------+--------------+
+|     Embedding     |     Word size    |      300      |      300      |                     300                    |                     300                    |      300     |
```
Comment (szha) [inline scripts/question_answering/index.rst]: Seems to have some formatting issue here: http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-411/7/model_zoo/question_answering/index.html

File: scripts/question_answering/index.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+~~~~~~~~~~~~~~~~~~~~
+To preprocess the data, you need to download `dataset <https://rajpurkar.github.io/SQuAD-explorer/>`_ and `glove <https://nlp.stanford.edu/projects/glove/>`_ pretrained word vector.
+And put all three files in the same directory with data_process.py and run the following code: 
+.. code-block:: console
+
+    $ python data_process.py
+
+[1] QANet_MXNet_1 (Val set F1= 79.0 EM= 69.8)
+
+Remember to modify the hyperparameter in config.py file to the value listed in the above table.
```
Comment (szha) [inline scripts/question_answering/index.rst]: Having preset config is great. On the other hand, instead of asking users to update the config file, it might be better to provide the presets beforehand so that users can switch with, say, a string argument to the script.

File: scripts/question_answering/config.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+parser.add_argument('--train_file_name', type=str,
+                    default='train-v1.1.json', help='The train file name.')
+parser.add_argument('--dev_file_name', type=str,
+                    default='dev-v1.1.json', help='The dev file name.')
+parser.add_argument('--processed_train_file_name', type=str,
+                    default='train_sorted.json', help='The processed train file name.')
+parser.add_argument('--processed_dev_file_name', type=str,
+                    default='dev_sorted.json', help='The processed dev file name.')
+parser.add_argument('--glove_file_name', type=str,
+                    default='glove.840B.300d.txt', help='The glove file name.')
```
Comment (szha) [inline scripts/question_answering/config.py]: This glove embedding is available in gluonnlp API with `gluonnlp.embedding.create('glove', 'glove.840B.300d')`, so that users don't have to download it manually.

File: scripts/question_answering/index.rst (hunk 1 above)
Comment (qyhfbqz) [reply scripts/question_answering/index.rst]: Fixed.

File: scripts/question_answering/index.rst (hunk 2 above)
Comment (qyhfbqz) [reply scripts/question_answering/index.rst]: Done.

**Comment by mli** [comment]
Job PR-411/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-411/8/index.html

File: scripts/question_answering/index.rst [hunk 4]
```diff
@@ -0,0 +1,68 @@
+Question Answer
+-------------------
+
```
Comment (szha) [inline scripts/question_answering/index.rst]: Could you add a download link here? GluonNLP's build system will automatically create a zip file for the scripts. You can use the following code to create a link for that file.
```
:download:`[Download] </model_zoo/question_answering.zip>`
```

Comment (qyhfbqz) [reply scripts/question_answering/index.rst]: Done.

File: scripts/question_answering/config.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+# pylint: disable=
+r"""
+This file contains all hyperparameters.
+"""
+import argparse
+
+import mxnet as mx
+
+parser = argparse.ArgumentParser(
+    description='The hyperparameters for QANet and optimizer.')
```
Comment (leezu) [inline scripts/question_answering/config.py]: Please specify `formatter_class=argparse.ArgumentDefaultsHelpFormatter`, so that the default values are included in the `--help` output

File: scripts/question_answering/train.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    model.collect_params().initialize(ctx=CTX)
+    # initial embedding parameters with glove
+    word_embedding = nd.array(load_emb_mat(opt.word_emb_file_name))
+
+    model.word_emb[0].weight.set_data(word_embedding)
+
+
+def reset_embedding_grad(model):
+    r"""
+    Reset the grad about word embedding layer.
```
Comment (leezu) [inline scripts/question_answering/train.py]: Does it mean you want to fix the embedding layer? If so, please disable gradients for the word embeddings instead of manually setting them to zero. That will save both GPU memory and computation

File: scripts/question_answering/model.py [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+                        Quoc V. Le},
+            title     = {QANet: Combining Local Convolution with Global Self-Attention for
+                        Reading Comprehension},
+            year      = {2018},
+            url       = {http://arxiv.org/abs/1804.09541}
+        }
+
+    """
+
+    def __init__(self, **kwargs):
```
Comment (leezu) [inline scripts/question_answering/model.py]: Please add keyword arguments for the options that you currently retrieve from `opt`, e.g. `opt.char_emb_dim`. The reason is, that if you require `opt`, QANet can only be used when people use the exact same `argparse` setup as in this script.

File: scripts/question_answering/model.py [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+    """
+
+    def __init__(self, dropout, **kwargs):
+        super(StochasticDropoutLayer, self).__init__(**kwargs)
+        self.dropout = dropout
+        with self.name_scope():
+            self.dropout_fn = gluon.nn.Dropout(dropout)
+
+    def hybrid_forward(self, F, inputs):
+        if F.random.uniform().asscalar() < self.dropout:
```
Comment (leezu) [inline scripts/question_answering/model.py]: I don't think you can do this in a HybridBlock? Did you try hybridizing the model? Does it work?

File: scripts/question_answering/model.py [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+    r"""
+    An implementation of position encoder.
+    """
+
+    def __init__(self, **kwargs):
+        super(PositionEncoder, self).__init__(**kwargs)
+        with self.name_scope():
+            pass
+
+    def hybrid_forward(self, F, x, min_timescale=1.0, max_timescale=1e4):
```
Comment (leezu) [inline scripts/question_answering/model.py]: I think setting python parameters in a `hybrid_forward` will not work. Did you try changing them? I believe the model will always use the value of the parameter used during the first forward pass. Therefore, please move `min_timescale=1.0, max_timescale=1e4` to the `__init__` function in case they need not be changed between different training iterations, or pass a 1D NDArray containing a single element instead of the Python floats

File: scripts/question_answering/model.py [hunk 10]
```diff
@@ -0,0 +1,752 @@
...
+        context_max_len : int
+        context_max_len : int
+
+        Returns
+        --------
+        similarity_mat : NDArray
+            output tensor with shape `(batch_size, context_sequence_length, query_sequence_length)`
+        """
+
+        subres0 = nd.tile(self.w4c(context), [1, 1, query_max_len])
```
Comment (leezu) [inline scripts/question_answering/model.py]: This does not work in a HybridBlock

**Comment by szha** [comment]
@zhiheng-huang you may be interested in this implementation.
